rclpy/src/rclpy/qos_events.cpp

```diff
@@ -0,0 +1,351 @@
+    static_cast<int>(data->requested_incompatible_qos.last_policy_kind));
+}
+
+py::object
+_offered_deadline_missed_to_py_object(const qos_event_callback_data_t * data)
+{
+  py::module qos_events = py::module::import(PYMODULE_NAME);
+  py::object pyclass = qos_events.attr("QoSOfferedDeadlineMissedInfo");
+  return pyclass(
+    data->offered_deadline_missed.total_count,
+    data->offered_deadline_missed.total_count_change);
+}
+
+py::object
+_liveliness_lost_to_py_object(const qos_event_callback_data_t * data)
+{
+  py::module qos_events = py::module::import(PYMODULE_NAME);
+  py::object pyclass = qos_events.attr("QoSLivelinessLostInfo");
+  return pyclass(
+    data->liveliness_lost.total_count,
+    data->liveliness_lost.total_count_change);
+}
+
+py::object
+_offered_incompatible_qos_to_py_object(const qos_event_callback_data_t * data)
+{
+  py::module qos_events = py::module::import(PYMODULE_NAME);
+  py::object pyclass = qos_events.attr("QoSOfferedIncompatibleQoSInfo");
+  return pyclass(
+    data->offered_incompatible_qos.total_count,
+    data->offered_incompatible_qos.total_count_change,
+    // enum does not specify underlying data type, need intermediate cast
+    static_cast<int>(data->offered_incompatible_qos.last_policy_kind));
+}
+
+qos_event_data_filler_function *
+qos_event_data_filler_function_for(py::capsule pyparent, py::object pyevent_type)
+{
+  if (strcmp(pyparent.name(), "rclpy_subscription_t") == 0) {
+    switch (rcl_subscription_event_type_from(pyevent_type)) {
+      case RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED:
+        return &_requested_deadline_missed_to_py_object;
+      case RCL_SUBSCRIPTION_LIVELINESS_CHANGED:
+        return &_liveliness_changed_to_py_object;
+      case RCL_SUBSCRIPTION_MESSAGE_LOST:
+        return &_message_lost_to_py_object;
+      case RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS:
+        return &_requested_incompatible_qos_to_py_object;
+      default:
+        throw py::value_error("event type for subscriptions not understood");
+        // although this suggests a misalignment between C and Python interfaces
+    }
+  }
+  if (strcmp(pyparent.name(), "rclpy_publisher_t") == 0) {
+    switch (rcl_publisher_event_type_from(pyevent_type)) {
+      case RCL_PUBLISHER_OFFERED_DEADLINE_MISSED:
+        return &_offered_deadline_missed_to_py_object;
+      case RCL_PUBLISHER_LIVELINESS_LOST:
+        return &_liveliness_lost_to_py_object;
+      case RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS:
+        return &_offered_incompatible_qos_to_py_object;
+      default:
+        throw py::value_error("event type for publishers not understood");
+        // although this suggests a misalignment between C and Python interfaces
+    }
+  }
+  throw py::type_error("event parent is neither a publisher nor a subscription");
+}
+
+py::capsule
+event_wrap_in_capsule(unique_cstruct_ptr<rcl_event_t> event, py::capsule pyparent)
+{
+  rclpy_handle_t * event_handle = _rclpy_create_handle(
```

Comment:
Mind using this pattern with `rclpy_create_handle_capsule` instead of `_rclpy_create_handle`? I think it's a little easier to follow.

https://github.com/ros2/rclpy/blob/41f9c2fea1fc82e7383c53689fdb2c2f74e03d47/rclpy/src/rclpy/subscription.cpp#L106-L122

Reply:
Done in https://github.com/ros2/rclpy/pull/723/commits/2795b162bed6b8bb5b5d4dbbf6e8a435258702c4.